### PR TITLE
on ownCloud upgrade: upgrade all apps in order, load important ones

### PR DESCRIPTION
**Pull request by blizzz**
Fix for #12801  as suggested in https://github.com/owncloud/core/issues/12801#issuecomment-72690475

Did not test yet, open for discussion. We should make updating right this time and don't run into regressions. We were almost stumbling about a similar issue for OC 8, and at least one is still existing in OC 7.

What do you think @PVince81 @DeepDiver1975 @icewind1991 @LukasReschke

### CODE REVIEW

File: lib/private/updater.php [hunk 1]
```diff
@@ -291,13 +291,47 @@ private function includePreUpdate($appId) {
...
+	/**
+	 * upgrades all apps within a major ownCloud upgrade. Also loads "priority"
+	 * (types authentication, filesystem, logging, in that order) afterwards.
+	 *
+	 * @throws NeedsUpdateException
+	 */
 	protected function doAppUpgrade() {
 		$apps = \OC_App::getEnabledApps();
+		$priorityTypes = array('authentication', 'filesystem', 'logging');
+		$pseudoOtherType = 'other';
```
Comment (LukasReschke) [inline lib/private/updater.php]: :question:

Comment (blizzz) [inline lib/private/updater.php]: Any other app will go in that array in line 320

File: lib/private/updater.php [hunk 2]
```diff
@@ -291,13 +291,47 @@ private function includePreUpdate($appId) {
...
+		$stacks = array($pseudoOtherType => array());
 
 		foreach ($apps as $appId) {
-			if (\OC_App::shouldUpgrade($appId)) {
-				\OC_App::updateApp($appId);
-				$this->emit('\OC\Updater', 'appUpgrade', array($appId, \OC_App::getAppVersion($appId)));
+			$priorityType = false;
```
Comment (PVince81) [inline lib/private/updater.php]: In theory you could already check for `shouldUpgrade` here and call  `continue` if false, so you don't need to add them to the stacks. On the other hand the code looks cleaner like this: first sort and then upgrade. And it makes it possible to move this sorting logic to a separate method if needed.

**Comment by PVince81** [comment]
The code looks good. Curious to find out whether the LDAP upgrade you tested yesterday would work with this fix ?

Also, how critical is it to get this on OC 8 too ? There is some risk of regression so I guess 8.0.1 is ok to give more time for testing.

**Comment by PVince81** [comment]
If it works, can you also add a unit test ?
You can use the AppConfig mock approach that was used here to mock app config/types: https://github.com/owncloud/core/blob/master/tests/lib/app.php#L389

**Comment by blizzz** [comment]
> The code looks good. Curious to find out whether the LDAP upgrade you tested yesterday would work with this fix ?

Will do a test today. But it should, sharing is filesystem-type, so LDAP should be up to date and running when sharing's turn comes.

> Also, how critical is it to get this on OC 8 too ? There is some risk of regression so I guess 8.0.1 is ok to give more time for testing.

We should check whether there are any other update routines that require user or file access. If so, they need to be changed (like we did with encryption) or we need this. I prefer to have it for 8.0.1 earliest in order to have some time for testing.

**Comment by PVince81** [comment]
@blizzz doesn't seem to work.

I setup an OC 6 and enabled many apps (LDAP, calendar, contacts, antivirus, impress, text editor, etc)
Then I upgraded to stable7 + this patch applied.

The upgrade page would show "upgrading" with a few steps (but no apps) and then suddenly it goes back to the update page and I have to click again. It never finishes.

When I switched back to "stable7" the update went through and I saw calendar, contact apps being upgraded too.

Maybe something wrong with the foreach:
```
{"reqId":"54d3a5799573e","app":"core","message":"starting upgrade from 7.0.4.2 to 7.0.4.2","level":0,"time":"2015-02-05T17:16:41+00:00","method":"GET","url":"\/owncloud\/core\/ajax\/update.php?requesttoken=3478b391ec1d506c90a9"}
{"reqId":"54d3a5799573e","app":"PHP","message":"file_put_contents(\/srv\/www\/htdocs\/owncloud\/.htaccess): failed to open stream: Permission denied at \/srv\/www\/htdocs\/owncloud\/lib\/private\/setup.php#253","level":0,"time":"2015-02-05T17:16:41+00:00","method":"GET","url":"\/owncloud\/core\/ajax\/update.php?requesttoken=3478b391ec1d506c90a9"}
{"reqId":"54d3a5799573e","app":"PHP","message":"Invalid argument supplied for foreach() at \/srv\/www\/htdocs\/owncloud\/lib\/private\/updater.php#331","level":0,"time":"2015-02-05T17:16:46+00:00","method":"GET","url":"\/owncloud\/core\/ajax\/update.php?requesttoken=3478b391ec1d506c90a9"}
{"reqId":"54d3a5799573e","app":"user_ldap","message":"No group filter is specified, LDAP group feature will not be used.","level":1,"time":"2015-02-05T17:16:46+00:00","method":"GET","url":"\/owncloud\/core\/ajax\/update.php?requesttoken=3478b391ec1d506c90a9"}
{"reqId":"54d3a5799573e","app":"PHP","message":"Cannot modify header information - headers already sent at \/srv\/www\/htdocs\/owncloud\/lib\/private\/user\/session.php#240","level":0,"time":"2015-02-05T17:16:46+00:00","method":"GET","url":"\/owncloud\/core\/ajax\/update.php?requesttoken=3478b391ec1d506c90a9"}
{"reqId":"54d3a5799573e","app":"PHP","message":"Cannot modify header information - headers already sent at \/srv\/www\/htdocs\/owncloud\/lib\/private\/user\/session.php#241","level":0,"time":"2015-02-05T17:16:46+00:00","method":"GET","url":"\/owncloud\/core\/ajax\/update.php?requesttoken=3478b391ec1d506c90a9"}
{"reqId":"54d3a5799573e","app":"PHP","message":"Cannot modify header information - headers already sent at \/srv\/www\/htdocs\/owncloud\/lib\/private\/user\/session.php#242","level":0,"time":"2015-02-05T17:16:46+00:00","method":"GET","url":"\/owncloud\/core\/ajax\/update.php?requesttoken=3478b391ec1d506c90a9"}
{"reqId":"54d3a5799573e","app":"PHP","message":"Cannot modify header information - headers already sent at \/srv\/www\/htdocs\/owncloud\/lib\/private\/user\/session.php#245","level":0,"time":"2015-02-05T17:16:46+00:00","method":"GET","url":"\/owncloud\/core\/ajax\/update.php?requesttoken=3478b391ec1d506c90a9"}
{"reqId":"54d3a5799573e","app":"PHP","message":"Cannot modify header information - headers already sent at \/srv\/www\/htdocs\/owncloud\/lib\/private\/user\/session.php#246","level":0,"time":"2015-02-05T17:16:46+00:00","method":"GET","url":"\/owncloud\/core\/ajax\/update.php?requesttoken=3478b391ec1d506c90a9"}
{"reqId":"54d3a5799573e","app":"PHP","message":"Cannot modify header information - headers already sent at \/srv\/www\/htdocs\/owncloud\/lib\/private\/user\/session.php#247","level":0,"time":"2015-02-05T17:16:46+00:00","method":"GET","url":"\/owncloud\/core\/ajax\/update.php?requesttoken=3478b391ec1d506c90a9"}
```

**Comment by PVince81** [comment]
I debugged a bit and it seems that the "documents" app isn't upgrading.
Got it, the "other" stack is broken. Will push a quickfix.

**Comment by PVince81** [comment]
With the commit I just pushed the update to stable7 worked properly (I applied it there too). Shares and reshares are also still there, no suspicious storage entry.

- [ ] add unit tests

**Comment by PVince81** [comment]
after that I updated to this branch and it also worked.

Note that the orginal issue might not be reproducible on the OC 7 -> OC 8 upgrade path because the files_sharing code that bugged isn't triggered (that was the code to convert the "Shared" folder to a regular folder for all users)

**Comment by PVince81** [comment]
@schiesbn ^

**Comment by PVince81** [comment]
Fixes encryption key migration: https://github.com/owncloud/core/issues/13950

**Comment by PVince81** [comment]
So as discussed with @blizzz unit testing this is close to impossible due to the static classes.
And we need a backportable fix.

**Comment by PVince81** [comment]
So how do we proceed ? I'd say merge for now, as unit tests are not possible.
My tests were positive so far. I even did a local backport to OC 7 which seemed to run fine.

:+1:

**Comment by PVince81** [comment]
Maybe @jnfrmarks can also help testing

**Comment by blizzz** [comment]
Yes, I by accident posted to comments into the related issue instead of here: https://github.com/owncloud/core/issues/12801#issuecomment-73249810 https://github.com/owncloud/core/issues/12801#issuecomment-73264391 but basically @PVince81 summed it up.

And one more tester is needed.

**Comment by karlitschek** [comment]
@jnfrmarks is not available the next few days.

**Comment by PVince81** [comment]
@icewind1991 @MorrisJobke can you help testing ? Thanks.

**Comment by DeepDiver1975** [comment]
Can we please spend a few minutes on how to unit test this class?

It's critical class and has to be unit tested - THX

**Comment by PVince81** [comment]
Note: the original issue (lost files + shares) might not be reproducible on the OC 7->OC 8 migration path because the update.php for files_sharing doesn't contain the code to migrate away from the "Shared" folder any more. But upgrading from OC 6 -> OC 7 will trigger the issue.

But also, the "Shared" folder isn't the only issue. There was another case related to migrating share keys with encryption.

Basically, the problem is: if any app migration is trying to call `getUsers()` without this fix, they will miss the LDAP users, so these users will NOT get migrated properly. After this fix, it should work fine.

**Comment by PVince81** [comment]
@DeepDiver1975 more than a few minutes were spent by @blizzz to try and find a way to unit test this. Due to the staticness of the class, unit testing has been deemed difficult if not impossible.
@blizzz wanna give it another try ?

**Comment by PVince81** [comment]
also keep in mind that we don't want to make OC_App non-static here because the fix has to be backportable to OC 7.

**Comment by DeepDiver1975** [comment]
> also keep in mind that we don't want to make OC_App non-static here because the fix has to be backportable to OC 7.

well - we need a second pr for this anyhow ... unit testing this for master at least would be better then nothing

**Comment by PVince81** [comment]
Ok let me make the PR for stable7 and stable8 then.

**Comment by blizzz** [comment]
Splitting the method (as i have it stashed) and writing a non-static wrapper for \OC_App should do the trick. For anything below master. For master we could investigate on how to refactor OC_App.

**Comment by DeepDiver1975** [comment]
> For master we could investigate on how to refactor OC_App.

including an additional function/option to get alls enabled apps of a certain type - will easy up this code from ma understanding

**Comment by PVince81** [comment]
stable 7 PR: https://github.com/owncloud/core/pull/14040
stable 8 PR: https://github.com/owncloud/core/pull/14041

**Comment by blizzz** [comment]
Nice, parts are already in lib/private/app/apmanager.php

**Comment by karlitschek** [comment]
:+1:

**Comment by LukasReschke** [comment]
:+1:

**Comment by blizzz** [comment]
:-1: :-1: :-1: see https://github.com/owncloud/core/pull/13865#issuecomment-73681355

**Comment by PVince81** [comment]
Ah yes, the master version should come with unit tests.
The stable8/stable7 are exempt of that because adding unit tests needs refactoring. And we don't want to backport refactored stuff.

**Comment by PVince81** [comment]
- [ ] something to look at, in case it's related: https://github.com/owncloud/core/issues/14101

**Comment by blizzz** [comment]
So, yeah, WIP and if you touch one thing, two others are affected. So I was also about to fix constructors and switch to IConfig, as AppConfig dies. Unfortunately this needs #13938. Or i stick with AppConfig for now, but i prefer to switch thereto as i am working with that stuff already.

**Comment by MorrisJobke** [comment]
@blizzz Then first of all use IAppConfig, I will migrate all the remaining AppConfig usages to the new version. That shouldn't be a problem. So go for AppConfig, because the syntax will keep the same.

**Comment by blizzz** [comment]
k, thx

File: core/ajax/update.php [hunk 3]
```diff
@@ -12,6 +12,7 @@
 	$updater = new \OC\Updater(
 			\OC::$server->getHTTPHelper(),
 			\OC::$server->getAppConfig(),
+			\Oc::$server->getAppManager(),
```
Comment (LukasReschke) [inline core/ajax/update.php]: Uppercase OC?

**Comment by blizzz** [comment]
So far, this moves everything needed from OC_App to AppManager, making the sorting method testable. Now doAppUpgrade is left.

**Comment by DeepDiver1975** [comment]
scheduled for 8.1 - PR for 8.0.1 has been merged https://github.com/owncloud/core/pull/14041

**Comment by blizzz** [comment]
Is there a reason why https://github.com/owncloud/core/blob/master/lib/private/app.php#L1206 is not part of https://github.com/owncloud/core/blob/master/lib/private/app/infoparser.php ? @DeepDiver1975

**Comment by DeepDiver1975** [comment]
> Is there a reason why https://github.com/owncloud/core/blob/master/lib/private/app.php#L1206 is not part of https://github.com/owncloud/core/blob/master/lib/private/app/infoparser.php ? @DeepDiver1975

not really - most probably caused by the dev who introduced parseAppInfo @MorrisJobke afaik :speak_no_evil:

**Comment by MorrisJobke** [comment]
> not really - most probably caused by the dev who introduced parseAppInfo @MorrisJobke afaik :speak_no_evil:

Blame it on me. I have no idea, why I didn't realized this. Yes it needs to be moved over :) PR si welcome :)

**Comment by blizzz** [comment]
I will move it, too, no worries. Wanted to be sure it wasn't there for a reason.

**Comment by blizzz** [comment]
As discussed with @DeepDiver1975 this PR will be postponed, because the refactoring also conflicts with https://github.com/owncloud/core/pull/14282. Instead we will for now port the OC 8.0 fix (https://github.com/owncloud/core/pull/14041) to master, and eventually i will continue to work on it based on what's left from here.

**Comment by PVince81** [comment]
https://github.com/owncloud/core/pull/14282 was merged so it might be possible to continue this, if needed

**Comment by MorrisJobke** [comment]
@PVince81 @blizzz Is this still a valid PR or was is replace by all the other cross referenced PRs?

**Comment by blizzz** [comment]
It's WIP, i needed to let go of it for 8.1.

**Comment by MorrisJobke** [comment]
@blizzz Do we want to split stuff out of this? To get some of your work into core? Maybe in smaller pieces? Would be nice to not lose this.

**Comment by blizzz** [comment]
That was the problem – i only intended to change a little piece, but this is knit together, I am afraid this needs such a big and complete deed.

**Comment by DeepDiver1975** [comment]
@blizzz do we want to revive this one? :speak_no_evil:

**Comment by blizzz** [comment]
@DeepDiver1975 actually, with the overhaul of the updater, I don't know whether this is all still applicable.

**Comment by DeepDiver1975** [comment]
meanwhile we have the app manager in core - no idea if interfaces match up .... rebase is a pain.

Let's extract 1a6d61e305b26d452df7a8ababae9602c0650afb and see where we are going ...

**Comment by DeepDiver1975** [comment]
this is already in master -> close
